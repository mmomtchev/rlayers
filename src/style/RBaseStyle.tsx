import React, {JSX, PropsWithChildren} from 'react';

import {RContext, RContextType} from '../context';
import debug from '../debug';

/**
 * @propsfor RBaseStyle
 */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface RBaseStyleProps extends PropsWithChildren<unknown> {}

/**
 * An abstract class used as base for all Style components, not meant to be used directly
 */
export default class RBaseStyle<P extends RBaseStyleProps> extends React.PureComponent<
    P,
    Record<string, never>
> {
    static contextType = RContext;
    protected static classProps: string[] = [];
    classProps: string[];
    ol: unknown;
    context: RContextType;

    constructor(props: Readonly<P>) {
        super(props);
        this.ol = null;
    }

    /* istanbul ignore next */
    protected create(props: P): unknown {
        throw new Error('RBaseStyle is an abstract class');
    }

    protected refresh(prevProps?: P): void {
        debug('refreshStyle', this);
        if (!prevProps) return;
        for (const p of this.classProps) {
            const m = p.charAt(0).toUpperCase() + p.substring(1);
            if ((prevProps && prevProps[p]) !== this.props[p]) {
                if (this.ol['set' + m]) {
                    this.ol['set' + m](this.props[p]);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(
                        `Underlying OpenLayers object does not support updating of ${p} after object creation. ` +
                            'If you are using an anonymous constant array or object, ' +
                            'consider assigning its value to a constant and then passing the constant or ' +
                            'use React.useMemo() to avoid this warning and improve performance.'
                    );
                }
            }
        }
    }

    /* istanbul ignore next */
    protected set(ol: unknown): void {
        return;
    }

    componentDidMount(): void {
        this.set(this.ol);
    }

    componentDidUpdate(
        prevProps: Readonly<P>,
        prevState: Readonly<unknown>,
        snapshot: unknown
    ): void {
        if (prevProps !== this.props) this.refresh(prevProps);
    }

    componentWillUnmount(): void {
        this.set(null);
    }

    render(): JSX.Element {
        if (!this.context) throw new Error('A style property must be part of a style');
        if (this.ol === null) {
            this.ol = this.create(this.props);
        }
        return null;
    }
}
