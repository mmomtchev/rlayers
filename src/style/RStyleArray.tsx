import React, {JSX} from 'react';
import {createRoot} from 'react-dom/client';
import {Feature} from 'ol';
import Style from 'ol/style/Style';
import Geometry from 'ol/geom/Geometry';

import {RContext, RContextType} from '../context';
import {default as RStyle, RStyleProps} from './RStyle';
import debug from '../debug';

/** An array of RStyle, can be an RStyle of its own
 * - this represents the OpenLayers concept of a style array
 *
 * It replaces the references on its children, so individual elements
 * in the array cannot be referenced
 *
 * It doesn't support caching yet
 *
 * Every style in the array must be a static style and not a function
 *
 * Arrays of style functions are not supported by OpenLayers and won't
 * be supported rlayers either
 */
export default class RStyleArray extends RStyle {
    constructor(props: Readonly<RStyleProps>) {
        super(props);
        this.childRefs = [];
        if (props.render) this.ol = this.style;
        else this.ol = [];
    }

    style = (f: Feature<Geometry>, r: number): Style | Style[] => {
        if (this.props.render) {
            const element = this.props.render(f, r);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            React.Children.map((element.props as any).children, (child) => {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
                if (React.isValidElement(child) && (child.type as Function) !== RStyle)
                    throw new TypeError('An RStyleArray should contain only RStyle elements');
            });
            const styleArray = [];
            const reactElement = (
                <RContext.Provider value={{...this.context, styleArray}}>
                    {/* eslint-disable-next-line @typescript-eslint/no-explicit-any */}
                    {(element.props as any).children}
                </RContext.Provider>
            );
            createRoot(document.createElement('div')).render(reactElement);
            return styleArray as Style[];
        }
        return this.ol as Style[];
    };

    refresh(prevProps?: RStyleProps): void {
        super.refresh(prevProps);
    }

    render(): JSX.Element {
        React.Children.map(this.props.children, (child) => {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
            if (React.isValidElement(child) && (child.type as Function) !== RStyle)
                throw new TypeError('An RStyleArray should contain only RStyle elements');
        });
        if (!this.props.render)
            return (
                <div className='_rlayers_RStyleArray'>
                    <RContext.Provider value={{...this.context, styleArray: this.ol as Style[]}}>
                        {this.props.children}
                    </RContext.Provider>
                </div>
            );
        return <React.Fragment />;
    }
}
