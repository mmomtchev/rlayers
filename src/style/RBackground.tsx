import React, {JSX} from 'react';
import {Text, Style, Fill, Stroke} from 'ol/style';

import {RContext} from '../context';
import {default as RBaseStyle, RBaseStyleProps} from './RBaseStyle';
import debug from '../debug';

/**
 * @propsfor RBackground
 */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface RBackgroundProps extends RBaseStyleProps {}

type Background = {
    setFill: (f: Fill) => void;
    setStroke: (s: Stroke) => void;
};

/**
 * Background element of a text
 *
 * Requires an `RText` context
 *
 * Provides an `RStyle` context - for `Fill` or `Stroke`
 */
export default class RBackground extends RBaseStyle<RBackgroundProps> {
    protected static classProps = [];
    ol: Background;

    protected create(props: RBackgroundProps): Background {
        this.classProps = RBackground.classProps;
        const parent = this.context.style as unknown as Text;
        if (!parent.setBackgroundFill || !parent.setBackgroundStroke)
            /* istanbul ignore next */
            throw new Error('Parent element does not support a background');
        return {
            setFill: parent.setBackgroundFill.bind(parent),
            setStroke: parent.setBackgroundStroke.bind(parent)
        };
    }

    render(): JSX.Element {
        super.render();
        return (
            <div className='_rlayers_RStyle_RBackground'>
                <RContext.Provider value={{...this.context, style: this.ol as unknown as Style}}>
                    {this.props.children}
                </RContext.Provider>
            </div>
        );
    }
}
