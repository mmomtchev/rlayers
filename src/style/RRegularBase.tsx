import React, {JSX} from 'react';
import {Image, Fill, Stroke, Style} from 'ol/style';

import {RContext} from '../context';
import RImage, {RImageProps} from './RImage';
import debug from '../debug';

/**
 * @propsfor RRegularBase
 */
export interface RRegularBaseProps extends RImageProps {
    /** Radius of the polygon */
    radius?: number;
}

/** Abstract class */
export default class RRegularBase<P extends RRegularBaseProps> extends RImage<P> {
    protected static classProps = RImage.classProps.concat(['radius']);
    ol: Image;
    stroke: Stroke;
    fill: Fill;

    /* istanbul ignore next */
    protected create(props: P): Image {
        throw new Error('RImage is an abstract class');
    }

    protected setStroke(s: Stroke): void {
        /* This a sneaky way around OpenLayers not supporting
         * setStroke/setFill on RegulaRBaseStyle-derived classes */
        this.stroke = s;
        this.ol = this.create(this.props);
        super.set(this.ol);
    }

    protected setFill(f: Fill): void {
        /* This a sneaky way around OpenLayers not supporting
         * setStroke/setFill on RegulaRBaseStyle-derived classes */
        this.fill = f;
        this.ol = this.create(this.props);
        super.set(this.ol);
    }

    render(): JSX.Element {
        super.render();
        return (
            <div className='_rlayers_RStyle'>
                <RContext.Provider value={{...this.context, style: this as unknown as Style}}>
                    {this.props.children}
                </RContext.Provider>
            </div>
        );
    }
}
