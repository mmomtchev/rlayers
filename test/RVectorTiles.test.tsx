import React from 'react';
import {fireEvent, render} from '@testing-library/react';

import {MVT} from 'ol/format';
import {Style} from 'ol/style';
import {Feature} from 'ol';
import {RLayerVectorTile, RMap} from 'rlayers';
import {RStyle, RCircle, RStroke} from 'rlayers/style';
import * as common from './common';
import CircleStyle from 'ol/style/Circle';
import {ViewStateLayerStateExtent} from 'ol/View';

const props = {
    url: 'https://rlayers.meteo.guru/tiles/admin/{z}/{x}/{y}',
    style: common.styles.yellow,
    format: new MVT()
};

const propsFeature = {
    ...props,
    format: new MVT<Feature>({featureClass: Feature})
};

describe('<RLayerVectorTiles>', () => {
    it('should create a vector tile layer', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile>;
        const {container, unmount} = render(
            <RMap {...common.mapProps}>
                <RLayerVectorTile {...props} ref={ref} renderBuffer={250} />
            </RMap>
        );
        expect(ref.current?.source.getProjection()?.getCode()).toBe('EPSG:3857');
        expect(container.innerHTML).toMatchSnapshot();
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        expect((ref.current?.ol as any).renderBuffer_).toBe(250);
        unmount();
    });
    it('should create a vector tile layer w/ RenderFeatures', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile<Feature>>;
        const {container, unmount} = render(
            <RMap {...common.mapProps}>
                <RLayerVectorTile {...propsFeature} ref={ref} renderBuffer={250} />
            </RMap>
        );
        expect(ref.current?.source.getProjection()?.getCode()).toBe('EPSG:3857');
        expect(container.innerHTML).toMatchSnapshot();
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        expect((ref.current?.ol as any).renderBuffer_).toBe(250);
        unmount();
    });
    it('should throw an error without a Map', () => {
        // eslint-disable-next-line no-console
        const err = console.error;
        // eslint-disable-next-line no-console
        console.error = () => undefined;
        expect(() => render(<RLayerVectorTile {...props} />)).toThrow('must be part of');
        // eslint-disable-next-line no-console
        console.error = err;
    });
    it('should relay OpenLayers events to features', async () => {
        const mapEvents = ['Click', 'PointerMove'] as const;
        const handler = jest.fn();
        const handlers = mapEvents.reduce((ac, a) => ({...ac, ['on' + a]: handler}), {});
        const map = React.createRef<RMap>();
        const layer = React.createRef<RLayerVectorTile>();
        const {unmount} = render(
            <RMap ref={map} {...common.mapProps}>
                <RLayerVectorTile ref={layer} {...props} {...handlers} />
            </RMap>
        );
        common.installMapFeaturesInterceptors(map.current!.ol, [
            {pixel: [10, 10], layer: layer.current!.ol, feature: new Feature()}
        ]);

        for (const ev of mapEvents) {
            // This should trigger a callback
            map.current!.ol.dispatchEvent(common.createEvent(ev, map.current!.ol, [10, 10]));
            // This should not
            map.current!.ol.dispatchEvent(common.createEvent(ev, map.current!.ol, [20, 20]));
        }
        expect(handler).toHaveBeenCalledTimes(mapEvents.length);
        unmount();
    });
    it('should generate enter/leave events to features ', async () => {
        const mapEvents = ['onPointerEnter', 'onPointerLeave'] as const;
        const handlers = mapEvents.reduce((ac, a) => ({...ac, [a]: jest.fn()}), {}) as Record<
            (typeof mapEvents)[number],
            () => void
        >;
        const map = React.createRef<RMap>();
        const layer = React.createRef<RLayerVectorTile>();
        const {container, unmount} = render(
            <RMap ref={map} {...common.mapProps}>
                <RLayerVectorTile ref={layer} {...props} {...handlers} />
            </RMap>
        );
        common.installMapFeaturesInterceptors(map.current!.ol, [
            {pixel: [10, 10], layer: layer.current!.ol, feature: new Feature()},
            {pixel: [20, 20], layer: layer.current!.ol, feature: new Feature()}
        ]);

        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(0);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(0);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [0, 0]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(0);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(0);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [10, 10]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(1);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(0);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [20, 20]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(2);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(1);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [0, 0]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(2);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(2);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [10, 10]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(3);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(2);

        map.current!.ol.dispatchEvent(common.createEvent('pointermove', map.current!.ol, [0, 0]));
        expect(handlers.onPointerEnter).toHaveBeenCalledTimes(3);
        expect(handlers.onPointerLeave).toHaveBeenCalledTimes(3);

        unmount();
    });
    it('should handle rerenders', async () => {
        const comp = (
            <RMap {...common.mapProps}>
                <RLayerVectorTile {...props} />
            </RMap>
        );
        const {rerender, container, unmount} = render(comp);
        expect(container.innerHTML).toMatchSnapshot();
        rerender(comp);
        expect(container.innerHTML).toMatchSnapshot();
        unmount();
    });
    it('should support projections', async () => {
        const layer = React.createRef() as React.RefObject<RLayerVectorTile>;
        const comp = (
            <RMap {...common.mapProps}>
                <RLayerVectorTile projection='EPSG:4326' ref={layer} {...props} />
            </RMap>
        );
        const {unmount} = render(comp);
        expect(layer.current?.source.getProjection()?.getCode()).toBe('EPSG:4326');
        unmount();
    });
    it('should support maxResolution / minResolution', async () => {
        const layer = React.createRef() as React.RefObject<RLayerVectorTile>;
        const comp = (
            <RMap {...common.mapProps}>
                <RLayerVectorTile
                    projection='EPSG:4326'
                    ref={layer}
                    {...props}
                    maxResolution={1000}
                    minResolution={100}
                />
            </RMap>
        );
        const {unmount, rerender} = render(comp);
        expect(layer.current?.ol.getMaxResolution()).toBe(1000);
        expect(layer.current?.ol.getMinResolution()).toBe(100);

        rerender(
            <RMap {...common.mapProps}>
                <RLayerVectorTile projection='EPSG:4326' ref={layer} {...props} />
            </RMap>
        );
        expect(layer.current?.ol.getMaxResolution()).toBeUndefined();
        expect(layer.current?.ol.getMinResolution()).toBeUndefined();

        unmount();
    });
    it('should update the style', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile>;
        const comp = (style) => (
            <RMap {...common.mapProps}>
                <RLayerVectorTile ref={ref} {...{...props, style}} />
            </RMap>
        );
        const {rerender, container, unmount} = render(comp(common.styles.yellow));
        expect(container.innerHTML).toMatchSnapshot();
        expect((ref.current?.ol.getStyle() as Style)?.getStroke()?.getWidth()).toBe(4);
        rerender(comp(common.styles.blueDot));
        expect(container.innerHTML).toMatchSnapshot();
        expect((ref.current?.ol.getStyle() as Style)?.getStroke()?.getWidth()).toBe(2);
        unmount();
    });
    it('should update the url', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile>;
        const comp = (url) => (
            <RMap {...common.mapProps}>
                <RLayerVectorTile ref={ref} {...{...props, url}} />
            </RMap>
        );
        const {rerender, container, unmount} = render(comp('http://url1'));
        expect(container.innerHTML).toMatchSnapshot();
        expect((ref.current?.source.getUrls() || [])[0]).toEqual('http://url1');
        rerender(comp('http://url2'));
        expect(container.innerHTML).toMatchSnapshot();
        expect((ref.current?.source.getUrls() || [])[0]).toEqual('http://url2');
        unmount();
    });
    it('should use inline styles', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile>;
        const comp = (width: number) => (
            <RMap {...common.mapProps}>
                <RLayerVectorTile ref={ref} {...{...props, style: undefined}}>
                    <RStyle zIndex={2}>
                        <RCircle radius={3}>
                            <RStroke color='#000005' width={width} />
                        </RCircle>
                    </RStyle>
                </RLayerVectorTile>
            </RMap>
        );
        const {rerender, container, unmount} = render(comp(1));
        expect(container.innerHTML).toMatchSnapshot();
        expect(
            ((ref.current?.ol.getStyle() as Style).getImage() as CircleStyle)
                ?.getStroke()
                ?.getWidth()
        ).toBe(1);
        rerender(comp(2));
        expect(container.innerHTML).toMatchSnapshot();
        expect(
            ((ref.current?.ol.getStyle() as Style).getImage() as CircleStyle)
                ?.getStroke()
                ?.getWidth()
        ).toBe(2);
        unmount();
    });
    it('should create a vector tile layer', async () => {
        const ref = React.createRef() as React.RefObject<RLayerVectorTile>;
        const {container, unmount} = render(
            <RMap {...common.mapProps}>
                <RLayerVectorTile {...props} ref={ref} attributions='Attributed' />
            </RMap>
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(
            ref.current.source.getAttributions()!(
                undefined as unknown as ViewStateLayerStateExtent
            )[0]
        ).toBe('Attributed');
        unmount();
    });
});
