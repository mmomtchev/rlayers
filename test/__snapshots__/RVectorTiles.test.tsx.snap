// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RLayerVectorTiles> should create a vector tile layer 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should create a vector tile layer 2`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should create a vector tile layer w/ RenderFeatures 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should handle rerenders 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should handle rerenders 2`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should update the style 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should update the style 2`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should update the url 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should update the url 2`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should use inline styles 1`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"><div class="_rlayers_RStyle"><div class="_rlayers_RStyle"></div></div></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;

exports[`<RLayerVectorTiles> should use inline styles 2`] = `"<div style="width: 100px; height: 100px;"><div class="_rlayers_RLayerVectorTile"><div class="_rlayers_RStyle"><div class="_rlayers_RStyle"></div></div></div><div class="ol-viewport" style="position: relative; overflow: hidden; width: 100%; height: 100%;"><div style="position: absolute; width: 100%; height: 100%; z-index: 0;" class="ol-unselectable ol-layers"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer"></div><div style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;" class="ol-overlaycontainer-stopevent"><div style="pointer-events: auto;" class="ol-zoom ol-unselectable ol-control"><button class="ol-zoom-in" type="button" title="Zoom in">+</button><button class="ol-zoom-out" type="button" title="Zoom out">–</button></div><div style="pointer-events: auto;" class="ol-rotate ol-unselectable ol-control ol-hidden"><button class="ol-rotate-reset" type="button" title="Reset rotation"><span class="ol-compass">⇧</span></button></div><div style="pointer-events: auto;" class="ol-attribution ol-unselectable ol-control ol-collapsed"><button type="button" aria-expanded="false" title="Attributions"><span class="ol-attribution-expand">i</span></button><ul></ul></div></div></div></div>"`;
